### PR TITLE
Blazor forms updates

### LINKED ISSUE #12476: Page is missing documentation how to make manual validation with OnSubmit
[Enter feedback here]

There is mention that you can use OnSubmit instead of OnValidSubmit. How to make custom validation (programmable)  and run data annotation validations there?
---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: eef47457-c8cf-74a4-b884-c1fe5c2ef52d
* Version Independent ID: 09a02d10-2834-467d-d8c2-3fd19ceb3863
* Content: [Blazor forms and validation](https://docs.microsoft.com/en-us/aspnet/core/blazor/forms-validation?view=aspnetcore-3.0)
* Content Source: [aspnetcore/blazor/forms-validation.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/blazor/forms-validation.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-blazor**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

### LINKED ISSUE #12617: Please describe how to bind submit button disabled state to validation IsValid
Given a /Page with an EditForm as shown in these docs, how would one bind the submit button's "disabled" attribute to the current validation status of the form?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: eef47457-c8cf-74a4-b884-c1fe5c2ef52d
* Version Independent ID: 09a02d10-2834-467d-d8c2-3fd19ceb3863
* Content: [Blazor forms and validation](https://docs.microsoft.com/en-us/aspnet/core/blazor/forms-validation?view=aspnetcore-3.0)
* Content Source: [aspnetcore/blazor/forms-validation.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/blazor/forms-validation.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-blazor**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

### LINKED ISSUE #15508: Input radio
Not sure what the right place for this is, but provide a description of how to work with input type=radio based on the comment in this bug https://github.com/aspnet/AspNetCore/issues/5579#issuecomment-548061223

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: eef47457-c8cf-74a4-b884-c1fe5c2ef52d
* Version Independent ID: 09a02d10-2834-467d-d8c2-3fd19ceb3863
* Content: [ASP.NET Core Blazor forms and validation](https://docs.microsoft.com/en-us/aspnet/core/blazor/forms-validation?view=aspnetcore-3.0)
* Content Source: [aspnetcore/blazor/forms-validation.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/blazor/forms-validation.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-blazor**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

### LINKED ISSUE #15915: Compare validator doesn't show validation error
Hi, I have checked with Compare validator in viewmodel for Email field. It is not showing any error in validation summary. Can you please check that?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: eef47457-c8cf-74a4-b884-c1fe5c2ef52d
* Version Independent ID: 09a02d10-2834-467d-d8c2-3fd19ceb3863
* Content: [ASP.NET Core Blazor forms and validation](https://docs.microsoft.com/en-us/aspnet/core/blazor/forms-validation?view=aspnetcore-3.0)
* Content Source: [aspnetcore/blazor/forms-validation.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/blazor/forms-validation.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-blazor**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #15508
Fixes #12617
Fixes #12476
Fixes #15915
Addresses #16270
Addresses #16273

[Internal Review Topic](https://review.docs.microsoft.com/en-us/aspnet/core/blazor/forms-validation?view=aspnetcore-3.0&branch=pr-en-us-15876)

* While I'm at it here, let's compose the StarShip Example 🚀 with the controls inside of the `<label></label>` tags.
* Use Javier's radio button example.
* Use @PeterHimschoot's *Attempt 2* approach ([blog post](https://blogs.u2u.be/peter/post/disabling-the-submit-button-in-blazor-with-validation)) for enabling/disabling the submit button based on form validation. Thanks Peter for writing that up! 🎸 ... a case of virtual :beer: to you for that one, sir!
* DR, given that Peter's blog has more goodness in it that devs might want to see (his later examples), would you like to cross-link it in an *Additional Resources* section here?
* This also fleshes out the `OnSubmit` bit with a Javier example. ✨
* ~Version the `CompareAttribute` content from the experimental package because https://github.com/aspnet/AspNetCore/issues/10643 was fixed in 3.1.~ Dropping because the topic will drop 3.0 coverage now.

### CODE REVIEW

**Comment by guardrex** [comment]
Ping @danroth27 ... this isn't blocking if ur busy tho.

File: aspnetcore/blazor/forms-validation.md [hunk 1]
```diff
@@ -184,6 +243,33 @@ Create a component with the following markup, and use the component just as `Inp
         this, __value => CurrentValueAsString = __value, CurrentValueAsString)" />
 ```
 
+## Work with radio buttons
+
+Blazor doesn't provide a built-in scenario for handling radio button `<input>` elements. Radio button support is under consideration for a future Blazor release. The following workaround is recommended:
```
Comment (danroth27) [inline aspnetcore/blazor/forms-validation.md]:
```suggestion
When working with radio buttons in a form, data binding is handled differently because radio buttons are evaluated as a group. The value of each radio button is fixed, but the value of the radio button group is the value of the selected radio button. The example below shows how to handle data binding for a radio button group:
```

Comment (danroth27) [reply aspnetcore/blazor/forms-validation.md]: We should probably also add a link to this section from the data binding section of the Components article.

File: aspnetcore/blazor/forms-validation.md [hunk 2]
```diff
@@ -184,6 +243,33 @@ Create a component with the following markup, and use the component just as `Inp
...
+<p>You chose: @currentChoice</p>
+
+@code {
+    private enum Choices { Red, Green, Blue };
+    private Choices currentChoice = Choices.Red;
+}
+```
+
+> [!NOTE]
+> The workaround shown in this section doesn't fire the `OnFieldChanged` form event that built-in `Input` components support.
```
Comment (danroth27) [inline aspnetcore/blazor/forms-validation.md]: This is unrelated to data binding to radio buttons - the `OnFieldChanged` event is used for validation. You can still validate when using radio buttons. We don't provide you an out of the box `InputRadio` component, but you can implement one like this:

```razor
@using System.Globalization
@using System.ComponentModel.DataAnnotations
@typeparam TValue

@inherits InputBase<TValue>

<input @attributes="AdditionalAttributes" type="radio" value="@SelectedValue" checked="@(SelectedValue.Equals(Value))" @onchange="OnChange" />

@code {
    [Parameter]
    public TValue SelectedValue { get; set; }

    void OnChange(ChangeEventArgs args)
    {
        CurrentValueAsString = args.Value.ToString();
    }

    protected override bool TryParseValueFromString(string value, out TValue result, out string validationErrorMessage)
    {
        var success = BindConverter.TryConvertTo<TValue>(value, CultureInfo.CurrentCulture, out var parsedValue);
        if (success)
        {
            result = parsedValue;
            validationErrorMessage = null;
            return true;
        }
        else
        {
            result = default;
            validationErrorMessage = $"The {FieldIdentifier.FieldName} field is not valid.";
            return false;
        }
    }
}
```

And then here's how you use it:

```razor
<EditForm Model="model" OnValidSubmit="OnValidSubmit">
    <DataAnnotationsValidator />
    <ValidationSummary />
    @for (int i = 1; i <= 5; i++)
    {
        <label>
            <InputRadio name="yourColor" SelectedValue="i" @bind-Value="model.Rating" />
            @i
        </label>
    }
    <button type="submit">Submit</button>
</EditForm>

<p>You chose: @model.Rating</p>

@code {
    private Model model = new Model();

    void OnValidSubmit()
    {
        Console.WriteLine("valid");
    }

    public class Model
    {
        [Range(1, 5)]
        public int Rating { get; set; }
    }
}
```

File: aspnetcore/blazor/forms-validation.md [hunk 3]
```diff
@@ -236,16 +322,10 @@ private class MyCustomValidator : ValidationAttribute
 }
 ```
 
-::: moniker range=">= aspnetcore-3.1"
-
 ### Blazor data annotations validation package
 
 The [Microsoft.AspNetCore.Blazor.DataAnnotations.Validation](https://www.nuget.org/packages/Microsoft.AspNetCore.Blazor.DataAnnotations.Validation) is a package that fills validation experience gaps using the `DataAnnotationsValidator` component. The package is currently *experimental*.
 
-### [CompareProperty] attribute
```
Comment (danroth27) [inline aspnetcore/blazor/forms-validation.md]: While the using the `CompareAttribute` isn't as broken in 3.1 as it was in 3.0 it still has issues because it doesn't associated the validation result with a specific member. This can result in inconsistent behavior between field-level validation and when the entire model is validated on a submit. The `ComparePropertyAttribute` addresses these issues and should still be recommended.

File: aspnetcore/blazor/forms-validation.md [hunk 4]
```diff
@@ -296,12 +376,69 @@ public class ShipDescription
...
+    <ValidationSummary />
+
+    ...
+
+    <button type="submit" disabled="@formInvalid">Submit</button>
+</EditForm>
+
+@code {
+    private Starship starship = new Starship();
+    private bool formInvalid = true;
```
Comment (pranavkm) [inline aspnetcore/blazor/forms-validation.md]:
```suggestion
    private bool formValid;
```

Comment (pranavkm) [reply aspnetcore/blazor/forms-validation.md]: Personal preference, but it's easier to think in terms of positives

File: aspnetcore/blazor/forms-validation.md [hunk 5]
```diff
@@ -296,12 +376,69 @@ public class ShipDescription
...
+    private EditContext editContext;
+
+    protected override void OnInitialized()
+    {
+        editContext = new EditContext(starship);
+
+        editContext.OnFieldChanged += (sender, e) =>
```
Comment (pranavkm) [inline aspnetcore/blazor/forms-validation.md]:
```C#
editContext.OnValidationStateChanged += (_, __) =>
{
  formValid = !editContext.GetValidationMessages().Any();
  StateHasChanged();
};
```

Comment (danroth27) [reply aspnetcore/blazor/forms-validation.md]: This doesn't quite work - fields that haven't been touched yet, but are in an invalid state, don't have validation messages yet.

Comment (guardrex) [reply aspnetcore/blazor/forms-validation.md]: Is what I have good (for now)?

Comment (guardrex) [reply aspnetcore/blazor/forms-validation.md]: ... and I understand the `_, __` discards ... just saw those recently. Should I use those now (for the first time :tada:) for `sender` and `e`?

Comment (guardrex) [reply aspnetcore/blazor/forms-validation.md]: ... and one more ❓ ... we've been using ...

```cshtml
<EditForm Model="@model" OnValidSubmit="HandleValidSubmit">
```

`OnValidSubmit` is set to `HandleValidSubmit` throughout the topic, but would you prefer to guide readers to what you just showed ...

```cshtml
<EditForm Model="@model" OnValidSubmit="OnValidSubmit">
```

Comment (danroth27) [reply aspnetcore/blazor/forms-validation.md]: I don't have a preference on what the event handler methods are named. What you have is fine.

File: aspnetcore/blazor/forms-validation.md (hunk 4 above)
Comment (guardrex) [reply aspnetcore/blazor/forms-validation.md]: Because the attribute is `disabled`, I felt it was simpler as `formInvalid` ... i.e., if the form is invalid, it's true, and the form is disabled (`disabled="true"`).

**Comment by guardrex** [comment]
@danroth27 @pranavkm Updates made.

Still an open ❓ on `formValid` vs. `formInvalid`. I put my reasoning on the comment :point_up:. Which approach do u prefer?

@danroth27 Would you make a quick ruling on underscores for fields on https://github.com/aspnet/AspNetCore.Docs/issues/15853. I'd like to try and knock these inconsistencies out of the docs as we go.

File: aspnetcore/blazor/forms-validation.md [hunk 6]
```diff
@@ -184,6 +243,93 @@ Create a component with the following markup, and use the component just as `Inp
         this, __value => CurrentValueAsString = __value, CurrentValueAsString)" />
 ```
 
+## Work with radio buttons
+
+When working with radio buttons in a form, data binding is handled differently than other elements because radio buttons are evaluated as a group. The value of each radio button is fixed, but the value of the radio button group is the value of the selected radio button. The following example shows how to handle data binding for a radio button group:
+
+The framework doesn't currently provide an `InputRadio` component. The following example shows how to implement an `InputRadio` component:
```
Comment (danroth27) [inline aspnetcore/blazor/forms-validation.md]: You have two sentences ending in colons one right after the other here.

File: aspnetcore/blazor/forms-validation.md [hunk 7]
```diff
@@ -184,6 +243,93 @@ Create a component with the following markup, and use the component just as `Inp
         this, __value => CurrentValueAsString = __value, CurrentValueAsString)" />
 ```
 
+## Work with radio buttons
+
+When working with radio buttons in a form, data binding is handled differently than other elements because radio buttons are evaluated as a group. The value of each radio button is fixed, but the value of the radio button group is the value of the selected radio button. The following example shows how to handle data binding for a radio button group:
```
Comment (danroth27) [inline aspnetcore/blazor/forms-validation.md]:
```suggestion
When working with radio buttons in a form, data binding is handled differently than other elements because radio buttons are evaluated as a group. The value of each radio button is fixed, but the value of the radio button group is the value of the selected radio button. The following example shows how to handle data binding for a radio button group and add support for validation using a custom `InputRadio` component:
```

File: aspnetcore/blazor/forms-validation.md (hunk 6 above)
Comment (danroth27) [reply aspnetcore/blazor/forms-validation.md]: Maybe remove this sentence and use the suggestion above?

**Comment by guardrex** [comment]
@pranavkm Do you have anything else?
